RV-Monitor/Kernel_Safety_Claims_by_Runtime_Verification_Monitors.md

```diff
@@ -0,0 +1,542 @@
+
+----->
+
+# Kernel Safety Claims by Runtime Verification Monitors
+
+
+# Introduction
+
+The goal of this document is to define a comprehensive methodology to design an RV Monitor against a set of allocated safety requirements. \
+Instead we do not intend to prove the RV Monitor to be able to accommodate every single use case.
+
+
+### Why it is hard to qualify the Kernel to support associated safety claims
+
+The Linux Kernel is monolithic and provides thousands of functionalities in order to support a huge amount of applications across multiple industry domains and across many different target HW platforms.
+
+For this reason the Kernel is designed to be modular (see [MAINTAINERS](https://github.com/torvalds/linux/blob/master/MAINTAINERS)), however each Kernel element (subsystem or driver) provides many functionalities and all such elements can interact with each other in a very complex way. To understand how complex such interactions can be, fig. [1] in the Appendix shows the possible call tree in the context of ioctl between “FILESYSTEMS (VFS and infrastructure)” and the drivers or subsystems it communicates with. \
+ \
+As an example, if there is a safety requirement associated with ioctl() (e.g. ioctl shall be used to set the correct timeout in an external safety watchdog device), even if the operation is in principle very simple, the Kernel code providing the implementation must be qualified to make sure:
+
+
+
+1. The kernel code implementing the requirement is correct;
+2. There is no unintended functionality in the current implementation;
+3. There is no undetected interference from other Kernel code outside the current implementation.
+
+With respect to points 1. and 2. above we can classify Kernel code into two classes:
+
+
+
+1. Kernel code that functionally contributes to the allocated safety requirement
+2. Kernel code potentially invoked as part of the implementation but not contributing to the safety requirement (e.g. printk).
+
+With respect to point 3. above we can have:
+
+
+
+3. Unintended functionalities due to interferences from Kernel code outside the implementation associated with classes A. and B. above. 
+
+
+When it comes to claim that code as in A. and B. are able to meet point 1 and point 2, for a complex SW component we need to hierarchically break it down into SW units, for each of these do a safety analysis and derive CoUs and Derived Safety Requirements; this regardless of code being classified as A. or B above. This process already proved to be very intensive even considering an entire subsystem like a single SW Unit (e.g. [STPA(-like) inside the Kernel ](https://docs.google.com/document/u/0/d/1K_cQSS2KYDnJQ0B91Zvlxq9-35Cx8ntbXwVMQJ51rvY/edit)).
+
+Finally with respect to failure modes as in C it is well known that, as of today, there are no comprehensive architectural measures that can isolate Kernel implementing the allocated safety requirement from code outside the implementation associated with A. and B.
```

Comment:
I would remove "well know", as it might give the impression that it's not factual.

Reply:
agreed. I'll rewrite as "with respect to failure modes as in C, as of today, there are"

Reply:
ok